docs/ovnkube.1
.TH OVNKUBE "1" "Jan 2018" "ovn-kubernetes" "OVN-KUBERNETES User Commands"
.SH NAME
ovnkube \- ovn utility
.SH SYNOPSIS
.B ovnkube
[\fI\,OPTION\/\fR]... [\fI\,FILE\/\fR]...
.SH DESCRIPTION
.PP
The ovn utility used to confiugre ovn-kubernetes
.SH GLOBAL OPTIONS
.TP
\fB\--cluster-subnet\fR string
A comma separated set of IP subnets and the associated hostsubnetlengths to use for
the cluster (eg, "10.128.0.0/14/23,10.0.0.0/14/23").  Each entry is given in the form
IP address/subnet mask/hostsubnetlength, the hostsubnetlength is optional and if
unspecified defaults to 24. The hostsubnetlength defines how many IP addresses are
dedicated to each node. (default "11.11.0.0/16")
.TP
\fB\--k8s-service-cidr\fR value
A CIDR notation IP range from which k8s assigns service cluster IPs.
This should be the same as the one provided for kube-apiserver's
\fB\--service-cluster-ip-range\fR option.
.TP
\fB\--init-master\fR string
Initialize master that acts as a controller and watches pods/nodes/services/policies. Requires the hostname as argument.
.TP
\fB\--init-node\fR string
Initialize node, requires the name that node is registered with in kubernetes cluster.
.TP
\fB\--remove-node\fR string
Remove a node from the OVN cluster, requires the name that node is registered
with in kubernetes cluster.
.TP
\fB\--gateway-mode\fR string
On nodes, creates a gateway for traffic to exit the cluster. Set to one of
"shared", "spare", or "local". Shared mode shares the gateway interface with
the OVN logical network. Spare mode takes over the gateway interface for use
with the OVN logical network. Local mode creates a NAT-ed interface for use
with the OVN logical network.
.TP
\fB\--init-gateways\fR
DEPRECATED; use \fB\--gateway-mode\fR instead.
.TP
\fB\--gateway-interface\fR string
The interface in nodes that will be the gateway interface.  If none
specified, then the node's interface on which the default gateway is
configured will be used as the gateway interface. Only useful with
\fB\--gateway-spare-interface\fR
DEPRECATED; use \fB\--gateway-mode\fR instead.
\fB--init-gateways\fR.
.TP
\fB\--gateway-nexthop\fR string
The external default gateway which is used as a next hop by
OVN gateway. This is many times just the default gateway
of the node in question. If not specified, the default gateway
configured in the node is used. Only useful with \fB--init-gateways\fR.
.TP
\fB\--gateway-spare-interface\fR
If true, assumes that "gateway-interface" provided can be exclusively
used for the OVN gateway.  When true, only OVNrelated traffic can flow
through this interface.
.TP
\fB\--gateway-local\fR
DEPRECATED; use \fB\--gateway-mode\fR instead.
.TP
\fB\--nodeport\fR
Setup nodeport based entries in OVN gateways for ingress into the k8s cluster.
By default, it is disabled.
.TP
\fB\--config-file\fR string
Configuration file path.
.TP
\fB\--mtu\fR value
MTU value used for the overlay networks. (default: 0).
.TP
\fB\--conntrack-zone\fR value
For gateway nodes, the conntrack zone used for conntrack flow rules (default: 0).
.TP
\fB\--loglevel\fR int
Log verbosity and level: 5=debug, 4=info, 3=warn, 2=error, 1=fatal (default: 0).
.TP
\fB\--logfile\fR string
Logfile name (with path) for ovnkube to write to.
.TP
\fB\--pidfile\fR string
(Optional) file name that will contain the pid.
.TP
\fB\--cni-conf-dir\fR string
The CNI config directory in which to write the overlay CNI config file.
.TP
\fB\--cni-plugin\fR string
The name of the CNI plugin.
.TP
\fB\--k8s-kubeconfig\fR string
Absolute path to the kubeconfig file (not required if the --k8s-apiserver, --k8s-cacert, and --k8s-token are given).
.TP
\fB\--k8s-apiserver\fR string
URL to the Kubernetes apiserver (default "https://localhost:8443") (not required if --k8s-kubeconfig is given).
.TP
\fB\--k8s-cacert\fR string
The absolute path to the Kubernetes API CA certificate (not required if --k8s-kubeconfig is given).
.TP
\fB\--k8s-token\fR string
The Kubernetes API authentication token (not required if --k8s-kubeconfig is given).
.TP
\fB\--nb-address\fR string
IP address and port of the OVN northbound API (eg, ssl://1.2.3.4:6641). Leave empty to use a local unix socket.
.TP
\fB\--nb-client-privkey\fR string
Private key that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--nb-client-cert\fR string
Client certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--nb-client-cacert\fR string
CA certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-address\fR string
IP address and port of the OVN southbound database (eg, ssl://1.2.3.4:6642).  Leave empty to use a local unix socket.
.TP
\fB\--sb-client-privkey\fR string
Private key that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-client-cert\fR string
Client certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-client-cacert\fR string
CA certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--help\fR, \fB\-h\fR
Show help.
.TP
\fB\--version\fR, \fB\-v\fR
Print the version.

.SH "SEE ALSO"
.BR ovn-k8s-overlay (1),
.BR ovn-kube-util (1),
.BR ovn_k8s (5).

.PP
https://github.com/ovn-org/ovn-kubernetes


go-controller/cmd/ovnkube/ovnkube.go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"net"
	"os"
	"os/signal"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"text/tabwriter"
	"text/template"

	"github.com/sirupsen/logrus"
	"github.com/urfave/cli"

	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cluster"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn"
	util "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"

	kexec "k8s.io/utils/exec"
)

const (
	// CustomAppHelpTemplate helps in grouping options to ovnkube
	CustomAppHelpTemplate = `NAME:
   {{.Name}} - {{.Usage}}

USAGE:
   {{.HelpName}} [global options]

VERSION:
   {{.Version}}{{if .Description}}

DESCRIPTION:
   {{.Description}}{{end}}

COMMANDS:{{range .VisibleCategories}}{{if .Name}}

   {{.Name}}:{{end}}{{range .VisibleCommands}}
     {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{end}}

GLOBAL OPTIONS:{{range $title, $category := getFlagsByCategory}}
   {{upper $title}}
   {{range $index, $option := $category}}{{if $index}}
   {{end}}{{$option}}{{end}}
   {{end}}`
)

func getFlagsByCategory() map[string][]cli.Flag {
	m := map[string][]cli.Flag{}
	m["Generic Options"] = config.CommonFlags
	m["K8s-related Options"] = config.K8sFlags
	m["OVN Northbound DB Options"] = config.OvnNBFlags
	m["OVN Southbound DB Options"] = config.OvnSBFlags
	m["OVN Gateway Options"] = config.OVNGatewayFlags

	return m
}

// borrowed from cli packages' printHelpCustom()
func printOvnKubeHelp(out io.Writer, templ string, data interface{}, customFunc map[string]interface{}) {
	funcMap := template.FuncMap{
		"join":               strings.Join,
		"upper":              strings.ToUpper,
		"getFlagsByCategory": getFlagsByCategory,
	}
	if customFunc != nil {
		for key, value := range customFunc {
			funcMap[key] = value
		}
	}

	w := tabwriter.NewWriter(out, 1, 8, 2, ' ', 0)
	t := template.Must(template.New("help").Funcs(funcMap).Parse(templ))
	err := t.Execute(w, data)
	if err == nil {
		_ = w.Flush()
	}
}

func main() {
	cli.HelpPrinterCustom = printOvnKubeHelp
	c := cli.NewApp()
	c.Name = "ovnkube"
	c.Usage = "run ovnkube to start master, node, and gateway services"
	c.Version = config.Version
	c.CustomAppHelpTemplate = CustomAppHelpTemplate
	c.Flags = config.CommonFlags
	c.Flags = append(c.Flags, config.K8sFlags...)
	c.Flags = append(c.Flags, config.OvnNBFlags...)
	c.Flags = append(c.Flags, config.OvnSBFlags...)
	c.Flags = append(c.Flags, config.OVNGatewayFlags...)
	c.Action = func(c *cli.Context) error {
		return runOvnKube(c)
	}

	if err := c.Run(os.Args); err != nil {
		logrus.Fatal(err)
	}
}

func delPidfile(pidfile string) {
	if pidfile != "" {
		if _, err := os.Stat(pidfile); err == nil {
			if err := os.Remove(pidfile); err != nil {
				logrus.Errorf("%s delete failed: %v", pidfile, err)
			}
		}
	}
}

func setupPIDFile(pidfile string) error {
	c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-c
		delPidfile(pidfile)
		os.Exit(1)
	}()

	// need to test if already there
	_, err := os.Stat(pidfile)

	// Create if it doesn't exist, else exit with error
	if os.IsNotExist(err) {
		if err := ioutil.WriteFile(pidfile, []byte(fmt.Sprintf("%d", os.Getpid())), 0644); err != nil {
			logrus.Errorf("failed to write pidfile %s (%v). Ignoring..", pidfile, err)
		}
	} else {
		// get the pid and see if it exists
		pid, err := ioutil.ReadFile(pidfile)
		if err != nil {
			logrus.Errorf("pidfile %s exists but can't be read", pidfile)
			return err
		}
		_, err1 := os.Stat("/proc/" + string(pid[:]) + "/cmdline")
		if os.IsNotExist(err1) {
			// Left over pid from dead process
			if err := ioutil.WriteFile(pidfile, []byte(fmt.Sprintf("%d", os.Getpid())), 0644); err != nil {
				logrus.Errorf("failed to write pidfile %s (%v). Ignoring..", pidfile, err)
			}
		} else {
			logrus.Errorf("pidfile %s exists and ovnkube is running", pidfile)
			os.Exit(1)
		}
	}

	return nil
}

func runOvnKube(ctx *cli.Context) error {
	exec := kexec.New()
	_, err := config.InitConfig(ctx, exec, nil)
	if err != nil {
		return err
	}

	if err = util.SetExec(exec); err != nil {
		logrus.Errorf("Failed to initialize exec helper: %v", err)
		return err
	}

	pidfile := ctx.String("pidfile")
	if pidfile != "" {
		defer delPidfile(pidfile)
		err = setupPIDFile(pidfile)
		if err != nil {
			return err
		}
	}

	clientset, err := util.NewClientset(&config.Kubernetes)
	if err != nil {
		panic(err.Error())
	}

	// create factory and start the controllers asked for
	stopChan := make(chan struct{})
	factory, err := factory.NewWatchFactory(clientset, stopChan)
	if err != nil {
		panic(err.Error())
	}

	master := ctx.String("init-master")
	node := ctx.String("init-node")
	cleanupNode := ctx.String("cleanup-node")
	if cleanupNode != "" {
		if master != "" || node != "" {
			panic("Cannot specify cleanup-node together with 'init-node or 'init-master'.")
		}

		err = cluster.CleanupClusterNode(&kube.Kube{KClient: clientset}, cleanupNode)
		if err != nil {
			logrus.Errorf(err.Error())
			panic(err.Error())
		}
		return nil
	}

	if master != "" || node != "" {
		if master != "" {
			if runtime.GOOS == "windows" {
				panic("Windows is not supported as master node")
			}
			ovnMasterController := ovn.NewMasterController(clientset, factory)
			ovnMasterController.ClusterIPNet, err = parseClusterSubnetEntries(ctx.String("cluster-subnet"))
			if err != nil {
				panic(err.Error())
			}
			if err = ovnMasterController.StartClusterMaster(master, true); err != nil {
				logrus.Errorf(err.Error())
				panic(err.Error())
			}
		}

		if node != "" {
			if config.Kubernetes.Token == "" {
				panic("Cannot initialize node without service account 'token'. Please provide one with --k8s-token argument")
			}

			ovnNodeController := cluster.NewOvnNodeController(clientset, factory)
			ovnNodeController.ClusterIPNet, err = parseClusterSubnetEntries(ctx.String("cluster-subnet"))
			err := ovnNodeController.StartClusterNode(node)
			if err != nil {
				logrus.Errorf(err.Error())
				panic(err.Error())
			}
		}

		// run forever
		select {}
	}
	return fmt.Errorf("need to run ovnkube in either master and/or node mode")
}

// parseClusterSubnetEntries returns the parsed set of CIDRNetworkEntries passed by the user on the command line
// These entries define the clusters network space by specifying a set of CIDR and netmaskas the SDN can allocate
// addresses from.
func parseClusterSubnetEntries(clusterSubnetCmd string) ([]ovn.CIDRNetworkEntry, error) {
	var parsedClusterList []ovn.CIDRNetworkEntry

	clusterEntriesList := strings.Split(clusterSubnetCmd, ",")

	for _, clusterEntry := range clusterEntriesList {
		var parsedClusterEntry ovn.CIDRNetworkEntry

		splitClusterEntry := strings.Split(clusterEntry, "/")
		if len(splitClusterEntry) == 3 {
			tmp, err := strconv.ParseUint(splitClusterEntry[2], 10, 32)
			if err != nil {
				return nil, err
			}
			parsedClusterEntry.HostSubnetLength = uint32(tmp)
		} else if len(splitClusterEntry) == 2 {
			// the old hardcoded value for backwards compatability
			parsedClusterEntry.HostSubnetLength = 24
		} else {
			return nil, fmt.Errorf("cluster-cidr not formatted properly")
		}

		var err error
		_, parsedClusterEntry.CIDR, err = net.ParseCIDR(fmt.Sprintf("%s/%s", splitClusterEntry[0], splitClusterEntry[1]))
		if err != nil {
			return nil, err
		}

		//check to make sure that no cidrs overlap
		if cidrsOverlap(parsedClusterEntry.CIDR, parsedClusterList) {
			return nil, fmt.Errorf("CIDR %s overlaps with another cluster network CIDR", parsedClusterEntry.CIDR.String())
		}

		parsedClusterList = append(parsedClusterList, parsedClusterEntry)

	}

	return parsedClusterList, nil
}

//cidrsOverlap returns a true if the cidr range overlaps any in the list of cidr ranges
func cidrsOverlap(cidr *net.IPNet, cidrList []ovn.CIDRNetworkEntry) bool {

	for _, clusterEntry := range cidrList {
		if cidr.Contains(clusterEntry.CIDR.IP) || clusterEntry.CIDR.Contains(cidr.IP) {
			return true
		}
	}
	return false
}
